File: internal/opensse/executor/vertex.go
package executor

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

func init() {
	RegisterSpecialized("vertex", &VertexExecutor{
		Base: Base{
			Provider: "vertex",
			Client:   nil,
			Headers:  nil,
			BaseURL:  "",
			BaseURLs: nil,
		},
		partner: false,
	})
	RegisterSpecialized("vertex-partner", &VertexExecutor{
		Base: Base{
			Provider: "vertex-partner",
			Client:   nil,
			Headers:  nil,
			BaseURL:  "",
			BaseURLs: nil,
		},
		partner: true,
	})
}

// VertexExecutor handles requests for Google Cloud Vertex AI and partner models.
type VertexExecutor struct {
	Base
	partner bool
}

func (e *VertexExecutor) projectID(cred Credentials) string {
	trimmed := strings.TrimSpace(cred.APIKey)
	if trimmed == "" || !strings.HasPrefix(trimmed, "{") {
		return ""
	}

	var sa map[string]any
	if err := json.Unmarshal([]byte(cred.APIKey), &sa); err != nil {
		return ""
	}

	if p, ok := sa["project_id"].(string); ok {
		return p
	}

	if p, ok := sa["quota_project_id"].(string); ok {
		return p
	}

	return ""
}

func (e *VertexExecutor) buildURL(model string, stream bool, cred Credentials) string {
	project := e.projectID(cred)

	rawKey := ""
	if !strings.HasPrefix(strings.TrimSpace(cred.APIKey), "{") && cred.AccessToken == "" {
		rawKey = cred.APIKey
	}

	if e.partner {
		return e.buildPartnerURL(project, rawKey)
	}

	return e.buildGeminiURL(model, stream, project, rawKey)
}

func (e *VertexExecutor) buildPartnerURL(project, rawKey string) string {
	if project == "" {
		project = "unknown"
	}

	url := fmt.Sprintf("https://aiplatform.googleapis.com/v1/projects/%s/locations/global/endpoints/openapi/chat/completions", project)
	if rawKey != "" {
		url += "?key=" + rawKey
	}

	return url
}

func (e *VertexExecutor) buildGeminiURL(model string, stream bool, project, rawKey string) string {
	location := "us-central1"

	action := "generateContent"
	if stream {
		action = "streamGenerateContent?alt=sse"
	}

	var url string
	if project == "" {
		url = fmt.Sprintf("https://aiplatform.googleapis.com/v1/publishers/google/models/%s:%s", model, action)
	} else {
		url = fmt.Sprintf("https://%s-aiplatform.googleapis.com/v1/projects/%s/locations/%s/publishers/google/models/%s:%s",
			location, project, location, model, action)
	}

	if rawKey != "" {
		if strings.Contains(url, "?") {
			url += "&key=" + rawKey
		} else {
			url += "?key=" + rawKey
		}
	}

	return url
}

func (e *VertexExecutor) buildHeaders(cred Credentials, stream bool) http.Header {
	h := make(http.Header)
	h.Set("Content-Type", "application/json")

	tok := cred.AccessToken
	if tok != "" {
		h.Set("Authorization", "Bearer "+tok)
	}

	if stream {
		h.Set("Accept", "text/event-stream")
	}

	return h
}

// Execute executes Vertex AI requests.
func (e *VertexExecutor) Execute(ctx context.Context, cred Credentials, model string, body []byte, stream bool) (*Result, error) {
	var m map[string]any
	if err := json.Unmarshal(body, &m); err != nil {
		return nil, err
	}

	if e.partner {
		m["model"] = model
		m["stream"] = stream
	}

	payload, err := json.Marshal(m)
	if err != nil {
		return nil, err
	}

	url := e.buildURL(model, stream, cred)
	if base := strings.TrimRight(cred.BaseURL, "/"); base != "" && strings.Contains(base, "aiplatform") {
		url = base
	}

	return e.DoPOST(ctx, url, e.buildHeaders(cred, stream), payload)
}

File: open-sse/src/executors/vertex.rs
//! Vertex AI Executor
use anyhow::{Context, Result, anyhow};
use reqwest::Client;
use serde_json::Value;
use std::time::Duration;
use crate::executors::default::{UpstreamBody, UpstreamResponse};
use crate::providers::Provider;

pub async fn execute(
    _provider: &Provider,
    model: &str,
    body: Value,
    stream: bool,
    credentials: &Value,
) -> Result<UpstreamResponse> {
    let client = Client::builder()
        .use_rustls_tls()
        .connect_timeout(Duration::from_secs(30))
        .build()?;

    let project = credentials.pointer("/providerSpecificData/project")
        .or_else(|| credentials.get("projectId"))
        .or_else(|| credentials.get("project"))
        .and_then(Value::as_str)
        .unwrap_or("default");

    let location = credentials.pointer("/providerSpecificData/location")
        .or_else(|| credentials.get("location"))
        .and_then(Value::as_str)
        .unwrap_or("us-central1");

    let token = credentials.get("accessToken")
        .or_else(|| credentials.get("apiKey"))
        .and_then(Value::as_str)
        .unwrap_or("");

    let action = if stream { "streamGenerateContent?alt=sse" } else { "generateContent" };
    let url = format!("https://{location}-aiplatform.googleapis.com/v1/projects/{project}/locations/{location}/publishers/google/models/{model}:{action}");

    let mut req = client.post(&url)
        .json(&body)
        .header("Authorization", format!("Bearer {token}"))
        .header("Content-Type", "application/json");

    if stream {
        req = req.header("Accept", "text/event-stream");
    } else {
        req = req.timeout(Duration::from_secs(300));
    }

    let resp = req.send().await.context("vertex request")?;
    let status = resp.status().as_u16();

    if stream && status == 200 {
        use futures_util::StreamExt;
        let byte_stream = resp.bytes_stream();
        let mapped = byte_stream.map(|r| r.map_err(|e| anyhow!(e)));
        return Ok(UpstreamResponse {
            status,
            body: UpstreamBody::Sse(Box::pin(mapped)),
            url,
        });
    }

    let json: Value = resp.json().await.context("parse vertex json")?;
    Ok(UpstreamResponse {
        status,
        body: UpstreamBody::Json(json),
        url,
    })
}
